sdk/storage/azure_storage_blob/src/partitioned_transfer/upload.rs

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use azure_core::http::Body;
+use bytes::Bytes;
+
+#[cfg(not(target_arch = "wasm32"))]
+use azure_core::stream::SeekableStream;
+#[cfg(not(target_arch = "wasm32"))]
+use futures::StreamExt;
+
+#[cfg(not(target_arch = "wasm32"))]
+use crate::streams::partitioned_stream::PartitionedStream;
+
+use super::*;
+
+#[async_trait::async_trait]
+pub(crate) trait PartitionedUploadBehavior {
+    async fn transfer_oneshot(&self, content: Body) -> AzureResult<()>;
+    async fn transfer_partition(&self, offset: usize, content: Body) -> AzureResult<()>;
+    async fn initialize(&self, content_len: usize) -> AzureResult<()>;
+    async fn finalize(&self) -> AzureResult<()>;
+}
+
+pub(crate) async fn upload(
+    content: Body,
+    parallel: NonZero<usize>,
+    partition_size: NonZero<usize>,
+    client: &impl PartitionedUploadBehavior,
+) -> AzureResult<()> {
+    if content.len() <= partition_size.get() {
```

Comment:
super nit: Would we want to catch the 0-size content length here and early exit, or do we just let it trickle to `transfer_oneshot` and if we want that optimization, implement it at the trait implementation level?